polygeist/lib/polygeist/Passes/ParallelLICM.cpp

```diff
@@ -28,305 +31,365 @@ struct ParallelLICM : public ParallelLICMBase<ParallelLICM> {
 };
 } // namespace
 
-static bool canBeParallelHoisted(Operation *op, Operation *scope,
-                                 SmallPtrSetImpl<Operation *> &willBeMoved) {
-  // Helper to check whether an operation is loop invariant wrt. SSA properties.
-  auto definedOutside = [&](Value value) {
-    if (auto BA = value.dyn_cast<BlockArgument>())
-      if (willBeMoved.count(BA.getOwner()->getParentOp()))
-        return true;
-    auto *definingOp = value.getDefiningOp();
-    if ((definingOp && !!willBeMoved.count(definingOp)) ||
-        cast<LoopLikeOpInterface>(scope).isDefinedOutsideOfLoop(value))
-      return true;
-    return false;
-  };
-
-  // Check that dependencies are defined outside of loop.
-  if (!llvm::all_of(op->getOperands(), definedOutside))
-    return false;
-
-  if (auto memEffect = dyn_cast<MemoryEffectOpInterface>(op)) {
-    SmallVector<MemoryEffects::EffectInstance, 1> effects;
-    memEffect.getEffects(effects);
-
-    SmallVector<SideEffects::Resource *> readResources;
-    SmallVector<SideEffects::Resource *> writeResources;
-    SmallVector<SideEffects::Resource *> freeResources;
-    for (auto effect : effects) {
-      if (isa<MemoryEffects::Allocate>(effect.getEffect()))
-        return false;
-      if (isa<MemoryEffects::Read>(effect.getEffect()))
-        readResources.push_back(effect.getResource());
-      if (isa<MemoryEffects::Write>(effect.getEffect()))
-        writeResources.push_back(effect.getResource());
-      if (isa<MemoryEffects::Free>(effect.getEffect()))
-        freeResources.push_back(effect.getResource());
+/// Represents the memory effects associated with an operation.
+struct OperationMemoryEffects {
+  OperationMemoryEffects(Operation &op) {
+    if (op.getName().hasInterface<MemoryEffectOpInterface>()) {
+      SmallVector<MemoryEffects::EffectInstance, 1> effects;
+      cast<MemoryEffectOpInterface>(op).getEffects(effects);
```

Comment:
```suggestion
    if (auto meop = dyn_cast<MemoryEffectOpInterface>(op)) {
      SmallVector<MemoryEffects::EffectInstance, 1> effects;
      meop.getEffects(effects);
```

Reply:
The are equivalent but I can make this change.